src/main/java/com/pylon/spokestack/SpeechContext.java

```diff
@@ -113,43 +176,127 @@ public Throwable getError() {
     /**
      * raises an error with the speech context.
      * @param value the exception to attach
+     * @return this
      */
-    public void setError(Throwable value) {
+    public SpeechContext setError(Throwable value) {
         this.error = value;
+        return this;
+    }
+
+    /**
+     * @return the current trace message
+     */
+    public String getMessage() {
+        return this.message;
     }
 
     /**
      * resets the context to the default state.
+     * @return this
      */
-    public void reset() {
+    public SpeechContext reset() {
+        setSpeech(false);
         setActive(false);
         setTranscript("");
         setConfidence(0);
         setError(null);
+        this.message = null;
+        return this;
+    }
+
+    /**
+     * traces a debug level message.
+     * @param format trace message format string
+     * @param params trace message format parameters
+     * @return this
+     */
+    public SpeechContext traceDebug(String format, Object... params) {
+        return trace(TraceLevel.DEBUG, format, params);
+    }
+
+    /**
+     * traces a performance level message.
+     * @param format trace message format string
+     * @param params trace message format parameters
+     * @return this
+     */
+    public SpeechContext tracePerf(String format, Object... params) {
+        return trace(TraceLevel.PERF, format, params);
+    }
+
+    /**
+     * traces an informational level message.
+     * @param format trace message format string
+     * @param params trace message format parameters
+     * @return this
+     */
+    public SpeechContext traceInfo(String format, Object... params) {
+        return trace(TraceLevel.INFO, format, params);
+    }
+
+    /**
+     * raises a trace event.
+     * @param level tracing level
+     * @param format trace message format string
+     * @param params trace message format parameters
+     * @return this
+     */
+    public SpeechContext trace(
+            TraceLevel level,
+            String format,
+            Object... params) {
+        if (canTrace(level)) {
+            this.message = String.format(format, params);
+            dispatch(Event.TRACE);
+        }
+        return this;
+    }
+
+    /**
+     * indicates whether a message will be traced at a level.
+     * @param level tracing level
+     * @return true if tracing will occur, false otherwise
+     */
+    public boolean canTrace(TraceLevel level) {
+        return level.value() >= this.traceLevel;
     }
 
     /**
      * dispatches a speech event.
      * @param event the event to publish
+     * @return this
      */
-    public void dispatch(Event event) {
-        for (OnSpeechEventListener listener: this.listeners)
-            listener.onEvent(event, this);
+    public SpeechContext dispatch(Event event) {
+        for (OnSpeechEventListener listener: this.listeners) {
+            try {
```

Comment:
This just handles the case where an event handler throws while handling an event. Previously, this would have thrown out of the pipeline, which could have corrupted state in one of the stages. Now that we have tracing, we can just trace these errors so the app can see them.